homeassistant/components/media_player/androidtv.py

```diff
@@ -0,0 +1,413 @@
+        self._scan_interval = scan_interval
+
+        self._uri = host + ':' + port
+        self._client = AdbClient(host="127.0.0.1", port=5037)
+        self._device = self._client.device(self._uri)
+
+        self._name = name
+        self._available = None
+        self._volume = None
+        self._muted = None
+        self._state = None
+        self._app_id = None
+        self._app_name = None
+        self._device = None
+        _LOGGER.info("its all good man")
+
+    def _init_regular_updates(self):
+        """Schedule regular updates at the specified interval."""
+        track_time_interval(
+            self._hass, lambda now: self.update(), self._scan_interval)
+
+    def update(self):
+        """Get the latest details from the device."""
+        try:
+            devices = self._client.devices()
+        except RuntimeError:
+            # Can't connect to adb server
+            if self._available:
+                _LOGGER.error(
+                    "Can't reach adb server, is it running ?")
+                _LOGGER.warning(
+                    "Device {} became unavailable.".format(self._name))
+                self._available = False
+        else:
+            if any([self._uri in host.get_serial_no() for host in devices]):
+                # Device is connected through the ADB server
+                if not self._available:
+                    _LOGGER.info("Device {} reconnected.".format(self._name))
+                    self._available = True
+
+                power_output = self._device.shell('dumpsys power')
+                audio_output = self._device.shell('dumpsys audio')
+                win_output = self._device.shell('dumpsys window windows')
+
+                self._state = self.get_state(power_output, audio_output)
+                self._muted, self._device, self._volume = self.get_audio(
+                    audio_output)
+                self._app_id = self.get_app_id(win_output)
+                self._app_name = self.get_app_name(self._app_id)
+
+            elif self._available:
+                _LOGGER.error(
+                    "ADB server not connected to {}".format(self._name))
+                _LOGGER.warning(
+                    "Device {} became unavailable.".format(self._name))
+                self._available = False
+
+    def get_state(self, power_output, audio_output):
+        """Process sys outputs and return the device state."""
+        if 'Display Power: state=ON' not in power_output:
+            return STATE_OFF
+        elif 'started' in audio_output:
+            state = STATE_PLAYING
+        elif 'paused' in audio_output:
+            state = STATE_PAUSED
+        else:
+            state = STATE_IDLE
+
+        return state
+
+    def get_audio(self, audio_output):
+        """Process sys output and return the volume, muted state and device."""
+        import re
+        block_pattern = 'STREAM_MUSIC(.*?)- STREAM'
+        stream_block = re.findall(
```

Comment:
Please extract device specific data parsing into standalone library published on pypi.

Reply:
Apart from this everything you mentioned should be fixed now, not sure when I'll have the time to figure that out and do it properly though.
Thanks for your time!